File: libs/server/domain/src/domain/member/workspace-access.ts
import { DomainError } from '../error';

export const WORKSPACE_ROLES = ['owner', 'member', 'viewer'] as const;
export type WorkspaceRole = (typeof WORKSPACE_ROLES)[number];
export type WorkspacePermission = 'read' | 'write' | 'manage';

const ROLE_PERMISSIONS: Record<WorkspaceRole, readonly WorkspacePermission[]> =
  {
    owner: ['read', 'write', 'manage'],
    member: ['read', 'write'],
    viewer: ['read'],
  };

export function workspaceRole(
  value: string,
  defaultRole?: WorkspaceRole,
): WorkspaceRole {
  const normalized = value.trim().toLowerCase() || defaultRole;
  if (isWorkspaceRole(normalized)) return normalized;
  throw DomainError.validation(`unsupported workspace role: ${value}`);
}

export function workspaceRoleAllows(
  role: string,
  permission: WorkspacePermission,
): boolean {
  return isWorkspaceRole(role) && ROLE_PERMISSIONS[role].includes(permission);
}

function isWorkspaceRole(value: unknown): value is WorkspaceRole {
  return (
    typeof value === 'string' &&
    (WORKSPACE_ROLES as readonly string[]).includes(value)
  );
}

File: libs/server-java/domain/src/main/java/reengineering/ddd/evidence/domain/validation/WorkspaceAccess.java
package reengineering.ddd.evidence.domain.validation;

import java.util.Locale;
import reengineering.ddd.evidence.domain.DomainException;

public final class WorkspaceAccess {
  private WorkspaceAccess() {}

  public static String role(String value, Role defaultRole) {
    String normalized = value == null ? "" : value.trim().toLowerCase(Locale.ROOT);
    if (normalized.isEmpty() && defaultRole != null) return defaultRole.wireValue();
    try {
      return Role.valueOf(normalized.toUpperCase(Locale.ROOT)).wireValue();
    } catch (IllegalArgumentException error) {
      throw DomainException.validation("unsupported workspace role: " + value);
    }
  }

  public static boolean allows(String role, Permission permission) {
    String normalized;
    try {
      normalized = role(role, null);
    } catch (DomainException error) {
      return false;
    }
    return switch (Role.valueOf(normalized.toUpperCase(Locale.ROOT))) {
      case OWNER -> true;
      case MEMBER -> permission != Permission.MANAGE;
      case VIEWER -> permission == Permission.READ;
    };
  }

  public enum Permission {
    READ,
    WRITE,
    MANAGE
  }

  public enum Role {
    OWNER,
    MEMBER,
    VIEWER;

    public String wireValue() {
      return name().toLowerCase(Locale.ROOT);
    }
  }
}
